### ISSUE TITLE
Refactor Feature Extraction to Use Dataclasses for Type Safety

### ISSUE THREAD

**Issue by Acusick1** [Acusick1/odds#14]
**Goal**: Replace string-keyed dictionaries with typed dataclasses in feature extraction to eliminate runtime errors from typos, provide compile-time type safety, and improve maintainability. Use `Optional[float]` with NaN for conditional features to properly distinguish missing data from legitimate zeros, enabling better downstream ML modeling.

**Complexity**: 🟡 Medium

**Location**:
- `packages/odds-analytics/odds_analytics/feature_extraction.py` - Add dataclasses, refactor both extractors
- `packages/odds-analytics/odds_analytics/ml_strategy_example.py` - Update feature access patterns  
- `tests/unit/test_feature_extraction.py` - Update feature access in tests
- `tests/integration/test_feature_extraction_integration.py` - Update integration tests

**Background**:

Current architecture uses string-keyed dictionaries with fragility:
```python
# Current: No compile-time checking, silent failures
features["consensus_prob"] = value  # Typo → silent failure, model trains on zeros
features["sharp_porb"] = value      # Would silently fail

# Missing features filled with 0.0 → semantic ambiguity
features.get("retail_sharp_diff", 0.0)  # Is this "no retail books" or "diff is actually zero"?
```

**Why Optional + NaN is Better**:

Current approach (fill missing with 0.0):
- ❌ Semantic ambiguity: Can't distinguish "feature unavailable" from "calculated value is zero"
- ❌ ML model confusion: `retail_sharp_diff = 0.0` could mean "no retail books" OR "retail = sharp"
- ❌ Misleading for downstream analysis

Proposed approach (Optional fields → NaN in array):
- ✅ Semantically correct: `None`/`NaN` explicitly means "not available"
- ✅ ML-friendly: XGBoost/LightGBM handle NaN natively, scikit-learn has imputation
- ✅ Clear intent: Consumer knows feature wasn't computed vs computed to zero
- ✅ Better debugging: Easy to identify missing data patterns

**Implementation Approach**:

Create separate dataclass for each extractor with optional fields for conditional features:

```python
from dataclasses import dataclass, fields
import numpy as np

@dataclass
class TabularFeatures:
    """Features for snapshot-based models (XGBoost, Random Forest)."""
    
    # Always available - consensus features (required)
    avg_home_odds: float
    avg_away_odds: float
    std_home_odds: float
    std_away_odds: float
    home_consensus_prob: float
    away_consensus_prob: float
    consensus_prob: float
    opponent_consensus_prob: float
    
    # Always available - market features (required)
    num_bookmakers: float
    avg_market_hold: float
    std_market_hold: float
    
    # Always available - best odds features (required)
    best_home_odds: float
    worst_home_odds: float
    home_odds_range: float
    best_away_odds: float
    worst_away_odds: float
    away_odds_range: float
    best_available_odds: float
    odds_range: float
    best_available_decimal: float
    
    # Always available - team indicators (required)
    is_home_team: float
    is_away_team: float
    
    # Conditional - sharp bookmaker features (optional: None if no sharp book)
    sharp_home_prob: float | None = None
    sharp_away_prob: float | None = None
    sharp_market_hold: float | None = None
    sharp_prob: float | None = None
    opponent_sharp_prob: float | None = None
    
    # Conditional - retail vs sharp (optional: None if no retail or sharp books)
    retail_sharp_diff_home: float | None = None
    retail_sharp_diff_away: float | None = None
    
    def to_array(self) -> np.ndarray:
        """Convert to numpy array, replacing None with NaN."""
        return np.array([
            val if val is not None else np.nan
            for val in (getattr(self, f.name) for f in fields(self))
        ])
    
    @classmethod
    def get_feature_names(cls) -> list[str]:
        """Return ordered list of feature names."""
        return [f.name for f in fields(cls)]

@dataclass
class SequenceFeatures:
    """Features per timestep for sequence models (LSTM, Transformers)."""
    
    # Always available (required)
    american_odds: float
    decimal_odds: float
    implied_prob: float
    num_bookmakers: float
    odds_std: float
    hours_to_game: float
    time_of_day_sin: float
    time_of_day_cos: float
    
    # Conditional - line movement (optional: None for first timestep or no opening line)
    odds_change_from_prev: float | None = None
    odds_change_from_opening: float | None = None
    implied_prob_change_from_prev: float | None = None
    implied_prob_change_from_opening: float | None = None
    
    # Conditional - sharp bookmaker (optional: None if no sharp book)
    sharp_odds: float | None = None
    sharp_prob: float | None = None
    retail_sharp_diff: float | None = None
    
    def to_array(self) -> np.ndarray:
        """Convert to numpy array, replacing None with NaN."""
        return np.array([
            val if val is not None else np.nan
            for val in (getattr(self, f.name) for f in fields(self))
        ])
    
    @classmethod
    def get_feature_names(cls) -> list[str]:
        return [f.name for f in fields(cls)]
```

**Refactoring Pattern**:

Before (fragile):
```python
def extract_features(self, event, odds_data, outcome, market):
    features = {}
    features["consensus_prob"] = calculate_prob(odds)
    if sharp_odds:
        features["sharp_prob"] = calculate_sharp(odds)
    # Missing sharp_prob → filled with 0.0 later
    return features  # dict[str, float]
```

After (type-safe):
```python
def extract_features(self, event, odds_data, outcome, market):
    consensus_prob = calculate_prob(odds)
    sharp_prob = calculate_sharp(odds) if sharp_odds else None  # Explicit None
    
    return TabularFeatures(
        consensus_prob=consensus_prob,
        sharp_prob=sharp_prob,  # None → NaN in array
        # ... all other fields (IDE enforces completeness)
    )  # TabularFeatures dataclass
```

**ML Model Compatibility**:

Models handle NaN well:
- **XGBoost/LightGBM**: Native NaN support (treats as missing, splits accordingly)
- **Scikit-learn**: Use `SimpleImputer` or indicator features
- **Neural networks**: Masking layers or learned embeddings
- **Pandas**: `df.isna()` for analysis

**Constraints**:
- Use `float | None` for optional fields (not `Optional[float]` - modern syntax)
- Convert `None → np.nan` in `to_array()` method
- All required fields must be provided at construction
- No new dependencies (use stdlib `dataclasses`)
- Use `from dataclasses import dataclass, fields` (not attrs or pydantic)
- Maintain same array ordering for ML model compatibility
- All tests must pass after refactor

**Scope**:
- Add `TabularFeatures` dataclass with 30 fields (23 required, 7 optional)
- Add `SequenceFeatures` dataclass with 15 fields (8 required, 7 optional)
- Implement `to_array()` method with None → NaN conversion
- Implement `get_feature_names()` classmethod using `fields()`
- Refactor `TabularFeatureExtractor.extract_features()` to return `TabularFeatures`
- Refactor `SequenceFeatureExtractor._extract_timestep_features()` to return `SequenceFeatures`
- Update `create_feature_vector()` to call `features.to_array()`
- Remove manual `_feature_names` lists (derive from dataclass fields)
- Update all tests to use attribute access (`features.consensus_prob`) instead of dict keys
- Update tests that check `"key" in features` to use `hasattr(features, "key")`
- Update `ml_strategy_example.py` to handle dataclass features
- Add test verifying NaN handling in arrays for missing optional features

**Success Criteria**:
- [ ] `TabularFeatures` dataclass defined with 30 fields (23 required + 7 optional)
- [ ] `SequenceFeatures` dataclass defined with 15 fields (8 required + 7 optional)
- [ ] Both dataclasses have `to_array()` method that converts `None → np.nan`
- [ ] Both dataclasses have `get_feature_names()` classmethod
- [ ] `TabularFeatureExtractor.extract_features()` returns `TabularFeatures` instance
- [ ] `SequenceFeatureExtractor._extract_timestep_features()` returns `SequenceFeatures` instance
- [ ] Optional fields correctly set to `None` when conditions not met
- [ ] `to_array()` produces NaN for optional fields that are None
- [ ] All 40 feature extraction tests pass
- [ ] No manual `_feature_names` lists remain in extractor classes
- [ ] IDE autocomplete works for feature names (verify in manual testing)
- [ ] Ruff check passes with no new warnings
- [ ] Test coverage maintained at ≥98%
- [ ] New test verifies NaN appears in array for missing optional features

**Benefits After Refactoring**:
- ✅ Compile-time type safety - typos caught with AttributeError
- ✅ IDE autocomplete for all features
- ✅ Semantic correctness - missing data clearly distinguished from zero
- ✅ ML-friendly - proper NaN handling enables better modeling
- ✅ Self-documenting - type hints show which features are conditional
- ✅ Easier refactoring - rename field propagates automatically
- ✅ Validation at construction - missing required field raises error immediately
- ✅ Better debugging - clear error messages for typos

**Related Files**:
- `packages/odds-analytics/odds_analytics/feature_extraction.py` - Main implementation
- `packages/odds-analytics/odds_analytics/backtesting/models.py` - See `BacktestEvent` for dataclass example pattern
- `tests/unit/test_feature_extraction.py` - Current dict-based test patterns to update

**Notes**:
- Python dataclasses are zero-overhead at runtime
- Feature ordering preserved via `fields()` iteration (insertion order guaranteed Python 3.7+)
- NaN propagates through numpy operations correctly
- After refactoring, adding new features is safer (missing field = immediate error vs silent 0.0)
- Optional field defaults to `None` - must explicitly set or it's missing
- Consider `@dataclass(frozen=True)` for immutability if desired, but not required
